utils/dataset.py

```diff
@@ -131,6 +132,146 @@ def __call__(self, batch: List[Dict[str, torch.tensor]]) -> Dict[str, torch.tens
+        mel_len = mel.shape[-1]
+        return {'x': x, 'mel': mel, 'item_id': item_id,
+                'mel_len': mel_len, 'x_len': len(x)}
+
+    def __len__(self):
+        return len(self.metadata)
+
+
+class ForwardDataset(Dataset):
+
+    def __init__(self,
+                 path: Path,
+                 dataset_ids: List[str],
+                 text_dict: Dict[str, str],
+                 tokenizer: Tokenizer):
+        self.path = path
+        self.metadata = dataset_ids
+        self.text_dict = text_dict
+        self.tokenizer = tokenizer
+
+    def __getitem__(self, index: int) -> Dict[str, torch.tensor]:
+        item_id = self.metadata[index]
+        text = self.text_dict[item_id]
+        x = self.tokenizer(text)
+        mel = np.load(str(self.path/'mel'/f'{item_id}.npy'))
+        mel_len = mel.shape[-1]
+        dur = np.load(str(self.path/'alg'/f'{item_id}.npy'))
+        pitch = np.load(str(self.path/'phon_pitch'/f'{item_id}.npy'))
+        energy = np.load(str(self.path/'phon_energy'/f'{item_id}.npy'))
+        return {'x': x, 'mel': mel, 'item_id': item_id, 'x_len': len(x),
+                'mel_len': mel_len, 'dur': dur, 'pitch': pitch, 'energy': energy}
+
+    def __len__(self):
+        return len(self.metadata)
+
+
+class BinnedTacoDataLoader:
+    """
+    Special dataloader to allow tacotron inference on batches with equal input length per batch.
+    This is used to safely generate attention matrices for extracting phoneme durations as there
+    is no input padding needed.
+    """
+
+    def __init__(self,
+                 data_path: Path,
+                 dataset: List[Tuple[str, int]],
+                 max_batch_size: int = 8) -> None:
+        """
+        Initializes the dataloader.
+
+        Args:
+            data_path: Main path to training data (paths.data)
+            dataset: List of Tuples with (file_id, mel_length)
+            max_batch_size: Maximum allowed batch size.
+        """
+
+        tokenizer = Tokenizer()
+        file_id_text_lens = []
+        text_dict = unpickle_binary(data_path / 'text_dict.pkl')
+        for item_id, _ in dataset:
+            toks = tokenizer(text_dict[item_id])
+            file_id_text_lens.append((item_id, len(toks)))
+
+        file_id_text_lens.sort(key=lambda x: x[1])
+        dataset_ids = [file_id for file_id, _ in file_id_text_lens]
+        dataset_lens = [text_len for _, text_len in file_id_text_lens]
+        dataset_lens = np.array(dataset_lens, dtype=int)
+        consecutive_split_points = np.where(np.diff(dataset_lens, append=0, prepend=0) != 0)[0]
+        dataset_indices = list(range(len(dataset)))
+        all_batches = []
+
+        for a, b in zip(consecutive_split_points[:-1], consecutive_split_points[1:]):
+            big_batch = dataset_indices[a:b]
+            batches = list(batchify(big_batch, batch_size=max_batch_size))
+            all_batches.extend(batches)
+
+        Random(42).shuffle(all_batches)
```

Comment:
would be nice to move seed value to some constant or config entry

Reply:
Moved to a constant. IMO not necessary to allow the setting in a config as the shuffling is merely used for a better time estimation of the inference.